### ISSUE TITLE
HAProxy blocks websocket requests to non-websocket endpoints

### ISSUE THREAD

**Issue by mathias-ioki** [grafana/grafana#67272]
## What happened:

We want to enable Live streaming/tailing for our Loki datasource. As suggested [here](https://github.com/grafana/grafana/issues/22905#issuecomment-628442315), we added the `Connection` and `Upgrade` http headers to the Loki datasource.

But as soon, as we add it, all connections to loki, except for the websocket connections, are not working anymore. We have a HAProxy in front of Loki, to loadbalance the traffic between our Loki instances. HAProxy marks the requests as invalid, as soon as the headers are added on Grafana side and the endpoint isn't a websocket endpoint.

In other words: Live tailing is working, when we add these headers, but all other requests are dropped by HAProxy because of the invalid request. So for example it's not possible to retrieve the label list from Loki.

The invalid request received by HAProxy looks like that:

```
[26/Apr/2023:09:38:52.479] frontend loki (#2): invalid request
  backend <NONE> (#-1), server <NONE> (#-1), event #4, src 2a00:c320:1000:f::24:60372
  buffer starts at 0 (including 0 out), 15706 free,
  len 678, wraps at 16336, error at position 678
  H1 connection flags 0x00000000, H1 stream flags 0x00000810
  H1 msg state MSG_DONE(30), H1 msg flags 0x00009590
  H1 chunk len 0 bytes, H1 body len 0 bytes :

  00000  GET /loki/api/v1/labels?start=1682491132353000000&end=1682494732353000
  00070+ 000 HTTP/1.1\r\n
  00084  Host: loki\r\n
  00107  User-Agent: Grafana/9.4.7\r\n
  00134  Accept: application/json, text/plain, */*\r\n
  00177  Accept-Encoding: gzip, deflate, br\r\n
  00213  Accept-Language: de,en-US;q=0.9,en-GB;q=0.8,en;q=0.7\r\n
  00267  Connection: upgrade\r\n
  00288  Dnt: 1\r\n
  00296  Sec-Ch-Ua: "Chromium";v="112", "Not:A-Brand";v="99"\r\n
  00349  Sec-Ch-Ua-Mobile: ?0\r\n
  00371  Sec-Ch-Ua-Platform: "macOS"\r\n
  00400  Sec-Fetch-Dest: empty\r\n
  00423  Sec-Fetch-Mode: cors\r\n
  00445  Sec-Fetch-Site: same-origin\r\n
  00474  Upgrade: websocket\r\n
  00494  X-Datasource-Uid: abcdefgh\r\n
  00523  X-Forwarded-For: 89.145.13.4, 127.0.0.1\r\n
  00564  X-Grafana-Org-Id: 3\r\n
  00585  X-Grafana-Referer: \r\n
  00606  X-Plugin-Id: loki\r\n
  00625  X-Real-Ip: 89.145.13.4\r\n
  00649  X-Scope-Orgid: loki.tenant\r\n
  00676  \r\n
```

## What you expected to happen:

Websocket specific headers should bet set automatically for websocket connections, for example for the `/loki/api/v1/tail` endpoint of Loki.

## How to reproduce it (as minimally and precisely as possible):

Connect Grafana to Loki via HAProxy in front of Loki and add the websocket headers to the Loki datasource.

## Anything else we need to know?:

It works if we connect Grafana directly to Loki without HAProxy in the middle.

## Environment:
- Grafana version: 9.4.7
- Data source type & version: Loki 2.7.4
- OS Grafana is installed on: Ubuntu 20.04.6
- User OS & Browser: MacOS Ventura, latest Chrome and latest Firefox

**Comment by svennergr** [comment]
Hey @mathias-ioki,

I don't think adding the `Connection` and `Upgrade` headers to every request is the right solution, even if it works in some situations. 

There's also not much we can do from our side at it looks like a problem with HAProxy. However, https://www.haproxy.com/blog/websockets-load-balancing-with-haproxy/#simple-configuration, suggests to add another websocket backend configuration for those cases:

```
backend bk_web                      
  balance roundrobin
  server websrv1 192.168.10.11:8080 maxconn 10000 weight 10 cookie websrv1 check
  server websrv2 192.168.10.12:8080 maxconn 10000 weight 10 cookie websrv2 check
```

Please let me know if that helped.

**Comment by mathias-ioki** [comment]
Hey @svennergr, thank you for your fast reply. I already tried the mentioned example, but doesn't really help. The Problem is, that all request are already dropped by HAProxy, before they are processed and routed.

I guess the main problem is, that we don't have a way, to set the mentioned headers only for websocket connections in Grafana. The only way we see at the moment is to set the headers on HAProxy side for URL paths we already know, that they need websocket connections. But this is really hacky and I hope for a better solution :)

**Comment by svennergr** [comment]
> I guess the main problem is, that we don't have a way, to set the mentioned headers only for websocket connections in Grafana.

There's not the need for it since the `Connection` and `Upgrade` headers are sent for live-tailing requests, both in Grafana's frontend and in the backend. 

Is there any chance you can share your HAProxy configuration?

**Comment by mathias-ioki** [comment]
You are absolutely right. It works now without setting the headers. I'm pretty sure I tested it before and it didn't work. But anyway, now it works and the headers are automatically set by Grafana as you mentioned. Thanks for your help and time and sorry for the confusion!
